pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,375 @@
+or leaves a topic. This is provided by the the ambient peer discovery
+mechanism and nominally not part of the router. A standalone
+implementation would have to implement those control messages.
+
+### Message processing
+
+Upon receiving a message, the router first processes the payload of the message.
+If it contains a valid message that has not been previously seen, then
+it publishes the message:
+- it forwards the message to every peer in `peers.floodsub[topic]`, provided it's not
+  the source of the message
+- It forwards the message to every peer in `mesh[topic]`, provided it's not the
+  source of the message.
+
+After processing the payload, it then processes the control messages in the envelope:
+- On `GRAFT(topic)` it adds the peer to `mesh[topic]` if it is
+  subscribed to the topic. If it is not subscribed, it responds with a `PRUNE(topic)`
+  control message.
+- On `PRUNE(topic)` it removes the peer from `mesh[topic]`.
+- On `IHAVE(ids)` it checks the `seen` set and requests uknown messages with an `IWANT`
+   message.
+- On `IWANT(ids)` it forwards all request messages that are present in `mcache` to the
+   requesting peer.
+
+When the router publishes a message that originates from the router itself (at the
+application layer), then it proceeds similar to the payload reaction:
+- it forwards the message to every peer in `peers.floodsub[topic]`.
+- if it is subscribed to the topic, then it must have a set of peers in `mesh[topic]`,
+  to which the message is forwarded.
+- if it is not subscribed to the topic, it then forwards the message to
+  the peers in `fanout[topic]`. If this set is empty, it chooses `D` peers from
+  `peers.gossipsub[topic]` to become the new `fanout[topic]` peers and forwards
+  to them.
+
+### Heartbeat
+
+The router periodically runs a hearbeat procedure, which is
+responsible for maintaining the mesh, emitting gossip, and shifting
+the message cache.
+
+The `mesh` is maintained exactly as prescribed by `meshsub`:
+```
+for each topic in mesh:
+ if |mesh[topic]| < D_low:
+   select D - |mesh[topic]| peers from peers.gossipsub[topic] - mesh[topic]
+   for each new peer:
+     add peer to mesh[topic]
+     emit GRAFT(topic) control message to peer
+
+ if |mesh[topic]| > D_high:
+   select |mesh[topic]| - D peers from mesh[topic]
+   for each new peer:
+     remove peer from mesh[topic]
+     emit PRUNE(topic) control message to peer
+```
+
+The `fanout` map is maintained by keeping track of last published time
+for each topic:
+```
+for each topic in fanout:
+ if last published time > TTL
+   remove topic from fanout
+ else if |fanout[topic]| < D
+   select D - |fanout[topic]| peers from peers.gossipsub[topic] - fanout[topic]
+   add the peers to fanout[topic]
+```
+
+Gossip is emitted by selecting peers for each topic that are not already part
+of the mesh:
+```
+for each topic in mesh+fanout:
+  let mids be mcache.window[topic]
+  if mids is not empty:
+    select D peers from peers.gossipsub[topic]
```

Comment:
Is there a reason why this `D` is the same as the target mesh degree `D`? I think it would make sense to be able to configure them separately because they fulfill different purposes: One might want to decrease the "eager D" (to use episub terminology) to save bandwidth, but increase the "lazy D" to still be safe (so that nodes still get all messages with high probability even if all of their few mesh peers drop out).

Reply:
There is no specific reason for using D, I just wanted to reduce the number of parameters.
But you are right, this can be different from D, and there is merit in having a separate parameter for this.

Reply:
I'll add some words to this effect, as this could be an important parameter.

Reply:
Add a paragraph about this.